File: parser.py
import click
import re
import requests
import json
from collections import OrderedDict


@click.command()
@click.option(
    '--url',
    default='https://raw.githubusercontent.com/1ppm/1ppmLog/master/HallOfFame.md',
    help='The URL to RAW content of MD file'
)
def parseHallOfFame(url):
    base_regex = r'\#\# \[([^]]+)\]\(([^)]+)\)([^##]+)'
    USERNAME_GROUP = 1
    USER_URL_GROUP = 2
    PROJECTS_GROUP = 3

    projects_regex = r'^\|([0-9\s\/]+)\|(.+?)\|(.+?)\|((.+)\|)?'
    PROJECT_DATE_GROUP = 1
    PROJECT_NAME_GROUP = 2
    PROJECT_DESCRIPTION_GROUP = 3
    PROJECT_STATUS_GROUP = 5

    data = []

    r = requests.get(url)
    content = r.text

    matches = re.finditer(base_regex, content)
    for matchNum, match in enumerate(matches):
        user_data = OrderedDict.fromkeys(['username', 'githubUrl', 'otherUrl', 'projects'], '')
        user_data['username'] = match.group(USERNAME_GROUP)
        user_data['projects'] = {}

        user_url = match.group(USER_URL_GROUP)
        if 'github' in user_url:
            user_data['githubUrl'] = user_url
        else:
            user_data['otherUrl'] = user_url

        projects_md = match.group(PROJECTS_GROUP)
        projects_matches = re.finditer(projects_regex, projects_md, re.MULTILINE)
        for projectMatchNum, projectMatch in enumerate(projects_matches):
            # Date processing
            project_date = ''.join(list(reversed(projectMatch.group(PROJECT_DATE_GROUP).strip().split('/'))))
            # Name processing
            project_name = projectMatch.group(PROJECT_NAME_GROUP).strip()
            parts = project_name.split(']')
            if len(parts) > 1:
                project_name = parts[0].lstrip('[')
            # Description processing
            project_description = projectMatch.group(PROJECT_DESCRIPTION_GROUP).strip()
            # Status processing
            project_status = ''
            if len(projectMatch.groups()) == PROJECT_STATUS_GROUP:
                project_status = projectMatch.group(PROJECT_STATUS_GROUP)
                project_status = project_status.strip() if project_status is not None else ''
            project_dict = OrderedDict.fromkeys(['name', 'description', 'status'], '')
            project_dict['name'] = project_name
            project_dict['description'] = project_description
            project_dict['status'] = project_status
            project = {
                project_date: project_dict
            }
            user_data['projects'].update(project)
        user_data['projects'] = OrderedDict(sorted(user_data['projects'].items(), key=lambda x: x[0]))
        data.append(user_data)
    data.sort(key=lambda d: d['username'])
    click.echo(json.dumps(data, indent=4))


if __name__ == '__main__':
    parseHallOfFame()

File: js_version/parser.js
var https = require('https');

var url = 'https://raw.githubusercontent.com/1ppm/1ppmLog/master/HallOfFame.md';

var baseRegex = /\#{1,2}\s\[([^\]]+)\]\(([^)]+)\)(\s\(.*(\@[^\)\]]+).*\))?(\n{1,}(\|.+\|))+/g;
var USERNAME_GROUP = 1;
var USER_URL_GROUP = 2;
var USER_TWITTER_GROUP = 4;
var PROJECTS_GROUP = 0;

var projectsRegex = /^\|([0-9\s\/]+)\|(.+?)\|(.+?)\|((.+)\|)?/gm;
var PROJECT_DATE_GROUP = 1;
var PROJECT_NAME_GROUP = 2;
var PROJECT_DESCRIPTION_GROUP = 3;
var PROJECT_STATUS_GROUP = 5;

var request = https.get(url, function(res) {
    var body = [];
    res.on('data', function(chunk) {
        body.push(chunk);
    }).on('end', function() {
        content = Buffer.concat(body).toString();
        var result = convertMDtoJSON(content);
        console.log(JSON.stringify(result, null, 4));
    });
});

function convertMDtoJSON(mdContent){
  var data = [];
  var usersMatch;
  while ((usersMatch = baseRegex.exec(mdContent)) !== null) {
    if (usersMatch.index === baseRegex.lastIndex) {
        baseRegex.lastIndex++;
    }
    
    var userData = {
        username: usersMatch[USERNAME_GROUP],
        githubUrl: '',
        twitterName: usersMatch[USER_TWITTER_GROUP] || '',
        otherUrl: '',
        projects: [] 
    };

    var userUrl = usersMatch[USER_URL_GROUP];
    if (userUrl.indexOf('github') > -1) {
        userData['githubUrl'] = userUrl;
    } else {
        userData['otherUrl'] = userUrl;
    }
    
    var projectsMD = usersMatch[PROJECTS_GROUP];
    var projectsMatch;
    while ((projectsMatch = projectsRegex.exec(projectsMD)) !== null) {
        if (projectsMatch.index === projectsRegex.lastIndex) {
            projectsRegex.lastIndex++;
        }
        // Date processing
        var projectDate = projectsMatch[PROJECT_DATE_GROUP].trim().split('/').reverse().join('');
        // Name processing
        var projectName = projectsMatch[PROJECT_NAME_GROUP].trim();
        var nameParts = projectName.split(']');
        if (nameParts.length) {
            projectName = nameParts[0].replace('[','');
        }
        // Description processing
        var projectDescription = projectsMatch[PROJECT_DESCRIPTION_GROUP].trim();
        var projectStatus = projectsMatch[PROJECT_STATUS_GROUP] ? projectsMatch[PROJECT_STATUS_GROUP].trim() : '';
        project = {
            date: projectDate,
            name: projectName,
            description: projectDescription,
            status: projectStatus
        }
        userData.projects.push(project);
    }
    data.push(userData);
  }
  return data;
}
